### ISSUE TITLE
Masks and weights in `weighted_objective`

### ISSUE THREAD

**Issue by linxihui** [fchollet/keras#3270]
At https://github.com/fchollet/keras/blob/master/keras/engine/training.py#L317

There is a division by `K.mean(mask)` and `K.mean(K.cast(K.not_equal(weights, 0), K.floatx()))`.  If batch size if small, one may have `mask` or `weights` all equal to `0`, which can lead to `NaN` issue. I ran into issue, and fixed it by wrapping them with `K.clip`.  Should this be fixed?

Here is the code of `weighted_objective`.

``` python
def weighted_objective(fn):
    '''Transforms an objective function `fn(y_true, y_pred)`
    into a sample-weighted, cost-masked objective function
    `fn(y_true, y_pred, weights, mask)`.
    '''
    def weighted(y_true, y_pred, weights, mask=None):
        # score_array has ndim >= 2
        score_array = fn(y_true, y_pred)
        if mask is not None:
            # Cast the mask to floatX to avoid float64 upcasting in theano
            mask = K.cast(mask, K.floatx())
            # mask should have the same shape as score_array
            score_array *= mask
            #  the loss per batch should be proportional
            #  to the number of unmasked samples.
            score_array /= K.mean(mask)

        # reduce score_array to same ndim as weight array
        ndim = K.ndim(score_array)
        weight_ndim = K.ndim(weights)
        score_array = K.mean(score_array, axis=list(range(weight_ndim, ndim)))

        # apply sample weighting
        if weights is not None:
            score_array *= weights
            score_array /= K.mean(K.cast(K.not_equal(weights, 0), K.floatx()))
        return K.mean(score_array)
return weighted
```

**Comment by fchollet** [comment]
Technically, it does not make sense to feed into your model a batch of data where *all* samples are masked. You should skip the batch, in your own code. That said, we might need protection against NaNs at this step regardless.

**Comment by linxihui** [comment]
If we have only one output node, yes we can just skip that batch. But if we have two output nodes (y1, y2), and y1 have weights all 0 while y2 has nonzero weights,  we don't want to skip this batch. But when computing the objective/score/loss for y1, NaNs occur and therefore the total/combined loss is NaN. But clip the mean, we can make the objective/score/loss for y1 = 0 instead of NaN.

**Comment by ragulpr** [comment]
@linxihui your fix is very useful/necessary but I'm not sure it's working or am I missing something. At present NaNs shows up when total mask or total weight is 0, both of which may happen randomly at sampling:

[Building on the @keras_test:](https://github.com/fchollet/keras/blob/master/tests/test_loss_masking.py)

    import numpy as np
    from keras.engine.training import _weighted_masked_objective
    from keras import losses
    from keras import backend as K

    weighted_loss = _weighted_masked_objective(losses.get('mae'))
    shape = (3, 4, 2)
    x = np.arange(24).reshape(shape)
    y = 2. * x

    # Normally the trailing 1 is added by standardize_weights
    weights = np.ones((3,))
    mask = np.ones((3, 4))
    # If we're in dead part of the mask:
    mask[:,:] = 0.
    weights[:] = 1.

    out = K.eval(weighted_loss(K.variable(x),
                               K.variable(y),
                               K.variable(weights),
                               K.variable(mask)))
    print(out)
    # > nan
    # .. Or the total sample weight is 0:
    mask[:,:] = 1.
    weights[:] = 0.

    out = K.eval(weighted_loss(K.variable(x),
                               K.variable(y),
                               K.variable(weights),
                               K.variable(mask)))
    print(out)
    # > nan

    import keras
    import tensorflow as tf
    print('version: '+keras.__version__) 
    #>version: 2.0.2
    print('version: '+tf.__version__) 
    #>version: 1.0.1

As a sidenote, I'm still not sure why/when mask and weights would be used in conjunction during training. It seems like it would always be cheaper to pass weights and infer mask from it than to test all inputs and weights at all time. That said, mask may still be useful during RNN-inference.
